### PR TITLE
Aspnet core windsor final castle windsor container extensions

### PR DESCRIPTION
**Pull request by generik0**
Hi @fir3pho3nixx 
I think we should expose the CrossWired though the options and as a extension on the container.

@fir3pho3nixx  I am very happey you have used a lot of the ideas i have provided, but it is a little demoralizing to see my work / ideas in the code (refactored by you) without a PR being merged/ appoved.

Please have a look. Any comments are welcome. Like the "WindsorContainerExtensions" name.
Let me have the commit :-) I know you have done 99.95% of the work and deserve 99.96% of the comits :-),

### CODE REVIEW

File: src/Castle.Facilities.AspNetCore/CastleWindsorRegistrationExtensions.cs [hunk 1]
```diff
@@ -237,37 +264,10 @@ public static void AddCastleWindsor(this IServiceCollection services, IWindsorCo
 		/// <param name="container">Windsor container which 'activators'/'cross wired' components resolve from</param>
 		/// <param name="serviceCollection">ASP.NET Core service collection from Microsoft.Extensions.DependencyInjection</param>
 		/// <param name="lifestyleType">The lifestyle type that needs to be copied across, only Singleton, Scoped and Transient are supported</param>
-		public static ComponentRegistration<TService> CrossWired<TService>(this ComponentRegistration<TService> registration, Type serviceType, IWindsorContainer container, IServiceCollection serviceCollection, LifestyleType lifestyleType = LifestyleType.Scoped) where TService : class
+		public static ComponentRegistration<object> CrossWired(this ComponentRegistration<object> registration, Type serviceType, IWindsorContainer container, IServiceCollection serviceCollection, LifestyleType lifestyleType = LifestyleType.Scoped)
```
Comment (fir3pho3nixx) [inline src/Castle.Facilities.AspNetCore/CastleWindsorRegistrationExtensions.cs]: Why was this generic parameter changed to object?

File: src/Castle.Facilities.AspNetCore/WindsorContainerExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+		/// <summary>
+		/// For making types available to the IServiceCollection using 'late bound' factories resolved via Windsor. This makes things like the @Inject directive in Razor work.
+		/// Service must be registered otherwise an ArgumentException will be thrown
+		/// </summary>
+		/// <param name="container">Windsor container which 'activators'/'cross wired' components resolve from</param>
+		/// <param name="serviceType">The type of the service</param>
+		/// <param name="serviceCollection">ASP.NET Core service collection from Microsoft.Extensions.DependencyInjection</param>
+		/// <param name="lifestyleType">The lifestyle type that needs to be copied across, only Singleton, Scoped and Transient are supported</param>
+		public static void CrossWired(this IWindsorContainer container, Type serviceType, IServiceCollection serviceCollection,
```
Comment (fir3pho3nixx) [inline src/Castle.Facilities.AspNetCore/WindsorContainerExtensions.cs]: Why are we exposing this on the container? It goes against how the fluent registration API works please see docs here: https://github.com/castleproject/Windsor/blob/master/docs/fluent-registration-api.md

Typically registration API's are surface via ComponentRegistration<T>.

**Comment by fir3pho3nixx** [comment]
I also liked the name `RegisterCrossWired` because it carries weight with the Castle Windsor community. They are used to types going into a container named this way via the public container API. The name means something more if you come from a Castle background. 

The name `CrossWired` leaves our community slightly in the dark as to what this method might be doing.

File: src/Castle.Facilities.AspNetCore/WindsorContainerExtensions.cs (hunk 2 above)
Comment (fir3pho3nixx) [reply src/Castle.Facilities.AspNetCore/WindsorContainerExtensions.cs]: At the very least we should rename this to `RegisterCrossWired`.

**Comment by generik0** [comment]
@fir3pho3nixx I can refactor in the morning if you like?
Let me know what needs to get done, and I will do it :-)

**Comment by fir3pho3nixx** [comment]
Yes, re-raise the PR, no problems :)

**Comment by fir3pho3nixx** [comment]
Please be aware you are also on a live PR at Castle: https://github.com/castleproject/Windsor/pull/394

If you have any doubts about the push, you can check our appveyor CI here: https://ci.appveyor.com/project/castleproject/windsor/build/0.0.377

File: src/Castle.Facilities.AspNetCore/WindsorContainerExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
+﻿using System;
```
Comment (fir3pho3nixx) [inline src/Castle.Facilities.AspNetCore/WindsorContainerExtensions.cs]: This file is missing the licensing header ...

**Comment by generik0** [comment]
OMG components not registered yet... They are being registered :-)

image

image

File: src/Castle.Facilities.AspNetCore/WindsorContainerExtensions.cs (hunk 2 above)
Comment (generik0) [reply src/Castle.Facilities.AspNetCore/WindsorContainerExtensions.cs]: Am i exposing the container? i am extending the container. So the owner will need to be in control of the container to do this. as mentioned below, this is to allow post registration cross wiring.

File: src/Castle.Facilities.AspNetCore/WindsorContainerExtensions.cs (hunk 3 above)
Comment (generik0) [reply src/Castle.Facilities.AspNetCore/WindsorContainerExtensions.cs]: fixed

File: src/Castle.Facilities.AspNetCore/WindsorContainerExtensions.cs (hunk 2 above)
Comment (fir3pho3nixx) [reply src/Castle.Facilities.AspNetCore/WindsorContainerExtensions.cs]: Recreating the ServiceProvider to support this comes with a severe performance penalty when it comes to memory management.

Comment (fir3pho3nixx) [reply src/Castle.Facilities.AspNetCore/WindsorContainerExtensions.cs]: I would rather throw a helpful exception which asks folks to register their `CrossWires` in the right place.

Comment (fir3pho3nixx) [reply src/Castle.Facilities.AspNetCore/WindsorContainerExtensions.cs]: Your choice I am happy to do both as long as our docs are in order.

**Comment by generik0** [comment]
@fir3pho3nixx I have fixed to your comments. 
However, do you have any good ideas as to InvalidOperationException with a custom message.
If i look at the service collections service descriptors all Implemenations are null.
So it looks like the ServiceCollection is not tracking anything (do we want this?), hence i cannot know if we are at startup and nothing has been resolved.
I could look into the container or something, but this will not indicate if the startup/ConfigureServices is done. I could look into the stack trace, but this is dirty.

Any ideas?

**Comment by fir3pho3nixx** [comment]
> However, do you have any good ideas as to InvalidOperationException with a custom message.
If i look at the service collections service descriptors all Implemenations are null. So it looks like the ServiceCollection is not tracking anything (do we want this?), hence i cannot know if we are at startup and nothing has been resolved.

This is less than ideal, let me pull down the branch and have a look. We might have to go the other way and reset the service provider. I am at work today but I will give this a closer look tonight.

**Comment by generik0** [comment]
@fir3pho3nixx Fyi [aspnet home issue on tracking](https://github.com/aspnet/Home/issues/3059#issuecomment-381527354)

**Comment by generik0** [comment]
@fir3pho3nixx how about I extent the servicecollection with crosswired instead of the container.
That way it usage is automatically preferred @configuretion..??

**Comment by fir3pho3nixx** [comment]
@generik0 

We should have chatted today

**Pull request merged**